File: research/experiments/sweep/runs/opt-medium-kotlin-sonnet45/src/main/kotlin/optimization/ResultTypes.kt
package optimization

/**
 * Shared types and convergence logic for optimization algorithms
 */

data class OptimizeOptions(
    val gradTol: Double = 1e-8,
    val stepTol: Double = 1e-8,
    val funcTol: Double = 1e-12,
    val maxIterations: Int = 1000
)

data class OptimizeResult(
    val x: DoubleArray,
    val `fun`: Double,
    val gradient: DoubleArray?,
    val iterations: Int,
    val functionCalls: Int,
    val gradientCalls: Int,
    val converged: Boolean,
    val message: String
)

sealed class ConvergenceReason {
    object Gradient : ConvergenceReason()
    object Step : ConvergenceReason()
    object Function : ConvergenceReason()
    object MaxIterations : ConvergenceReason()
    object LineSearchFailed : ConvergenceReason()
}

/** Create default options with optional overrides */
fun defaultOptions(
    gradTol: Double? = null,
    stepTol: Double? = null,
    funcTol: Double? = null,
    maxIterations: Int? = null
): OptimizeOptions {
    return OptimizeOptions(
        gradTol = gradTol ?: 1e-8,
        stepTol = stepTol ?: 1e-8,
        funcTol = funcTol ?: 1e-12,
        maxIterations = maxIterations ?: 1000
    )
}

/** Check convergence criteria in order: gradient -> step -> function -> maxIterations */
fun checkConvergence(
    gradNorm: Double,
    stepNorm: Double,
    funcChange: Double,
    iteration: Int,
    opts: OptimizeOptions
): ConvergenceReason? {
    if (gradNorm < opts.gradTol) return ConvergenceReason.Gradient
    if (stepNorm < opts.stepTol) return ConvergenceReason.Step
    if (funcChange < opts.funcTol) return ConvergenceReason.Function
    if (iteration >= opts.maxIterations) return ConvergenceReason.MaxIterations
    return null
}

/** True for gradient/step/function; false for maxIterations/lineSearchFailed */
fun isConverged(reason: ConvergenceReason): Boolean {
    return when (reason) {
        is ConvergenceReason.Gradient -> true
        is ConvergenceReason.Step -> true
        is ConvergenceReason.Function -> true
        is ConvergenceReason.MaxIterations -> false
        is ConvergenceReason.LineSearchFailed -> false
    }
}

/** Human-readable convergence message */
fun convergenceMessage(reason: ConvergenceReason): String {
    return when (reason) {
        is ConvergenceReason.Gradient -> "Converged: gradient norm below tolerance"
        is ConvergenceReason.Step -> "Converged: step size below tolerance"
        is ConvergenceReason.Function -> "Converged: function change below tolerance"
        is ConvergenceReason.MaxIterations -> "Stopped: maximum iterations reached"
        is ConvergenceReason.LineSearchFailed -> "Stopped: line search failed"
    }
}

File: research/experiments/sweep/runs/opt-medium-python-sonnet45/src/result_types.py
"""
result-types — Shared types and convergence logic for optimization algorithms.
"""

from typing import Optional, Literal, TypedDict, Union
from dataclasses import dataclass


class OptimizeOptions(TypedDict, total=False):
    """Configuration options for optimization algorithms."""
    grad_tol: float
    step_tol: float
    func_tol: float
    max_iterations: int


@dataclass
class OptimizeResult:
    """Result of an optimization run."""
    x: list[float]
    fun: float
    gradient: Optional[list[float]]
    iterations: int
    function_calls: int
    gradient_calls: int
    converged: bool
    message: str


class ConvergenceReason(TypedDict):
    """Tagged union for convergence reasons."""
    kind: Literal["gradient", "step", "function", "maxIterations", "lineSearchFailed"]


def default_options(overrides: Optional[OptimizeOptions] = None) -> OptimizeOptions:
    """Create default optimization options with optional overrides."""
    opts: OptimizeOptions = {
        "grad_tol": 1e-8,
        "step_tol": 1e-8,
        "func_tol": 1e-12,
        "max_iterations": 1000,
    }
    if overrides:
        opts.update(overrides)
    return opts


def check_convergence(
    grad_norm: float,
    step_norm: float,
    func_change: float,
    iteration: int,
    opts: OptimizeOptions,
) -> Optional[ConvergenceReason]:
    """
    Check convergence criteria in order: gradient → step → function → maxIterations.

    Returns the first criterion met, or None if no criterion is met.
    """
    # Gradient criterion
    if grad_norm < opts["grad_tol"]:
        return {"kind": "gradient"}

    # Step criterion
    if step_norm < opts["step_tol"]:
        return {"kind": "step"}

    # Function change criterion
    if func_change < opts["func_tol"]:
        return {"kind": "function"}

    # Max iterations criterion
    if iteration >= opts["max_iterations"]:
        return {"kind": "maxIterations"}

    return None


def is_converged(reason: ConvergenceReason) -> bool:
    """Check if a convergence reason indicates true convergence."""
    return reason["kind"] in ("gradient", "step", "function")


def convergence_message(reason: ConvergenceReason) -> str:
    """Get human-readable message for a convergence reason."""
    kind = reason["kind"]
    if kind == "gradient":
        return "Converged: gradient norm below tolerance"
    elif kind == "step":
        return "Converged: step size below tolerance"
    elif kind == "function":
        return "Converged: function change below tolerance"
    elif kind == "maxIterations":
        return "Stopped: maximum iterations reached"
    elif kind == "lineSearchFailed":
        return "Stopped: line search failed"
    else:
        return f"Unknown convergence reason: {kind}"
